client_plugin/drivers/shared/etcdops.go

```diff
@@ -269,8 +308,143 @@ func etcdService(cmd []string) {
 	}
 }
 
-func (d *VolumeDriver) createEtcdClient() *etcdClient.Client {
-	dclient := d.dockerd
+// checkLocalEtcd function check if local ETCD endpoint is successfully started or not
+// if yes, start the watcher for volume global refcount
+func (e *etcdInfo) checkLocalEtcd() error {
+	ticker := time.NewTicker(checkSleepDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(requestTimeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Infof("Checking etcd client is started")
+			cli, err := addrToEtcdClient(e.nodeAddr)
+			if err != nil {
+				log.WithFields(
+					log.Fields{"nodeAddr": e.nodeAddr,
+						"error": err},
+				).Warningf("Failed to get etcd client, retry before timeout ")
+			} else {
+				e.client = cli
+				go e.etcdWatcher()
+				return nil
+			}
+		case <-timer.C:
+			return fmt.Errorf("Timeout reached; etcd cluster is not started")
+		}
+	}
+}
+
+func (e *etcdInfo) etcdWatcher() {
+	// TODO: when the manager is demoted to worker, the watcher should be cancelled
+	watchCh := e.client.Watch(context.Background(), etcdPrefixGRef,
+		etcdClient.WithPrefix(), etcdClient.WithPrevKV())
+	for wresp := range watchCh {
+		for _, ev := range wresp.Events {
+			e.etcdEventHandler(ev)
+		}
+	}
+}
+
+func (e *etcdInfo) etcdEventHandler(ev *etcdClient.Event) {
+	log.WithFields(
+		log.Fields{"type": ev.Type},
+	).Infof("Watcher on global refcount returns event ")
+
+	if ev.Type == etcdClient.EventTypePut {
+		if string(ev.Kv.Value) == etcdSingleRef &&
+			ev.PrevKv != nil &&
+			string(ev.PrevKv.Value) == etcdNoRef {
+			// watcher observes global refcount from 0 to 1
+			// transactional edit state first
```

Comment:
We **cannot** have those CompareAndPut as a single transaction :)
As Shivanshu has mentioned above, please consider the _CompareAndPut to change the state from Ready to Intermediate_ as "lock", and consider the _CompareAndPut to change state from Intermediate to Ready_ as "unlock". Only when the "lock" is successful, we go ahead to do corresponding operations (that is, startSMBServer), and "unlock" later.